### ISSUE TITLE
[False positive] Redis::set() parameters

### ISSUE THREAD

**Issue by sebsel** [nunomaduro/larastan#907]
- Larastan Version: v0.7.12
- `--level` used: 5

### Description

I get PHPStan errors on a piece of code that just seems to work:
```php
Redis::set($key, true, 'EX', $seconds);
```

gives:
```
- Parameter #3 $timeout of static method Redis::set() expects array|int, string given.
- Static method Redis::set() invoked with 4 parameters, 2-3 required.
```

The [Redis Connection](https://github.com/nunomaduro/larastan/blob/b2da312efe88d501aeeb867ba857e8c4198d43c0/stubs/Redis/Connection.stub) has a mixin of `\Redis`, while I think it should be `Illuminate\Redis\Connections\PhpRedisConnection`.

From the [PhpRedis documentation for the `\Redis@set()`](https://github.com/phpredis/phpredis#set) method:
```php
$redis->set('key', 'value', ['nx', 'ex' => 10]);
```
... this seems to be what PHPStan expects. But when I do this in Tinker:
```
>>> Illuminate\Support\Facades\Redis::set('key', 'value', ['nx', 'ex' => 10])
TypeError: Illegal offset type
```

... therefore, I don't think the Facade resolves to `\Redis`, but to `Illuminate\Redis\Connections\PhpRedisConnection`. This class has [the following `set()`](https://github.com/laravel/framework/blob/eaacad4ad2a8e4cecdb1d98a344e1d16206415f4/src/Illuminate/Redis/Connections/PhpRedisConnection.php#L83):

```php
    /**
     * Set the string value in the argument as the value of the key.
     *
     * @param  string  $key
     * @param  mixed  $value
     * @param  string|null  $expireResolution
     * @param  int|null  $expireTTL
     * @param  string|null  $flag
     * @return bool
     */
    public function set($key, $value, $expireResolution = null, $expireTTL = null, $flag = null)
```

And this works in Tinker (on PHP 8):

```
>>> Illuminate\Support\Facades\Redis::set('key', 'value', expireResolution: 'EX', expireTTL: 2)
=> true
```

Note that I'm using the default `phpredis` value in my `config('database.redis.client')`. This includes a `pecl install phpredis` to set it up. You can also do a `composer install predis/predis` and set the driver to `predis`. In this case, I can do the following in Tinker:
```
>>> Illuminate\Support\Facades\Redis::set('key', 'value', expireResolution: 'EX', expireTTL: 2)
=> Predis\Response\Status {#3920}
>>> Illuminate\Support\Facades\Redis::set('key', 'value', ['nx', 'ex' => 10])
TypeError: strlen(): Argument #1 ($string) must be of type string, array given
```

Either way, the third argument does not seem to accept an array.

I think choosing `Illuminate\Redis\Connections\PhpRedisConnection` as mixin is the better option :)

**Comment by szepeviktor** [comment]
Thank you for your thorough report!

> The [Redis Connection](https://github.com/nunomaduro/larastan/blob/b2da312efe88d501aeeb867ba857e8c4198d43c0/stubs/Redis/Connection.stub) has a mixin of `\Redis`, while I think it should be `Illuminate\Redis\Connections\PhpRedisConnection`.

Could you modify it and run Larastan again?

**Comment by sebsel** [comment]
Hm, good point!

Changing `stubs/Redis/Connection.stub` did not work, but then I found `stubs/Facades.stub`. When I change the `@mixin` for the Redis facade to `\Illuminate\Redis\Connections\PhpRedisConnection`, the error disappears!

**Comment by canvural** [comment]
What about `PredisConnection`? Are the method signature same?

**Comment by sebsel** [comment]
There is a lot of magic in `Illuminate\Redis\Connections\PredisConnection` and `Predis\Client`, so I am not sure. The `PredisConnection` seems to proxy to `Predis\Client` and I think you are right: it might not have the same signature. I can't seem to find much documentation for Predis to verify this.

It seems weird indeed to point this to `PhpRedisConnection` specificly, but there does not seem to be a common interface available. Not sure what to do about that.

**Comment by szepeviktor** [comment]
PHP Redis is the superior interface for Redis as it is written in C. Predis is like shim...

**Comment by szepeviktor** [comment]
You are right @sebsel. Please send a PR if the local solution is not fit for you.

**Comment by canvural** [comment]
This is one of the stuff that I don't like in Laravel. There is a `ConnectionInterface` but the classes that implement that interface can define their own public methods 🤷

So, I guess the solution can be to add both of them to mixin and hope everything goes fine? 😄

**Comment by sebsel** [comment]
I don't want to get into "Predis or PHP Redis, which one is superior?". The point is that the interface of the `PhpRedisConnection` seems to be written out, while the other one is implicit. From my tests, it seems to me the interfaces match (as in: the code still works the same way). The typehint of `\Redis` does not seem to be correct, for the code does not  behave in that way.

So I take it that 'add both' is adding `\Illuminate\Redis\Connections\PhpRedisConnection` and `\Illuminate\Redis\Connections\PredisConnection`, and removing `\Redis`?

**Comment by bernardwiesner** [comment]
Sorry to revive this, but I got a similar issue with `Redis::hexists(...` . I am using predis, and predis returns an int while phpredis returns bool, so when I do:

```
return Redis::hexists('foo', 'bar') === 1;
```

I get an error:

```
Strict comparison using === between bool and 1 will always evaluate to false.
```

For now I will just change from `===` => `==`

**Comment by mikethea1** [comment]
Is there a plan to fix this? I'm seeing the same issue with this call:
```php
Redis::set($key, $value, 'EX', $renewExpirySeconds, 'NX');
```

Also hit with:
```php
Redis::eval('...', 1, $key, self::KEY_TTL_SECONDS);
```
